### DISCUSSION TITLE
[Solved] running python script with poetry

### DISCUSSION THREAD

**Discussion by junefish** [community/community#34884] [Actions] [answered]
I'm trying to set up a Github action to automatically run a python poetry script.  The manual command for running it ([source repo](https://github.com/radeklat/todoist-habitica-sync)) is `poetry run python src/main.py`, but github actions says "unknown command poetry" when I tried this.

I found a ton of guides for how to install and setup poetry, but none that give an example command for running a script once that is done.  I tried `python -m poetry run python -m src/main.py` and got:

```
Run python -m poetry run python -m src/main.py
/home/runner/work/todoist-habitica-sync/todoist-habitica-sync/.venv/bin/python: Error while finding module specification for 'src/main.py' (ModuleNotFoundError: No module named 'src/main'). Try using 'src/main' instead of 'src/main.py' as the module name.
Error: Process completed with exit code 1.
```
So I tried using `src/main` instead of `src/main.py` instead and got this:
```
Run python -m poetry run python -m src/main
  
/home/runner/work/todoist-habitica-sync/todoist-habitica-sync/.venv/bin/python: No module named src/main
Error: Process completed with exit code 1.
```

I'm not sure what I'm missing or what other info would be helpful to provide.  TIA!

**Comment by airtower-luna** [comment]
If you want to provide the module name instead of the source file, you need to replace the directory separator with a dot, e.g `python -m src.main`.

**Reply by junefish** [reply]
I tried changing the run command, and got a different error, so...progress?

```

Run python -m poetry run python -m src.main
  
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/todoist-habitica-sync/todoist-habitica-sync/src/main.py", line 8, in <module>
    from config import TODOIST_PRIORITY_TO_HABITICA_DIFFICULTY, get_settings
ModuleNotFoundError: No module named 'config'
Error: Process completed with exit code 1.
```

Now I'm confused as to why it can't find the config module - again, it works if I clone the repo locally and do a manual run.

**Comment by antonblr** [comment, marked as answer]
@junefish - How do you install `poetry` in your action, could you provide a GitHub action workflow snippet. I'd recommend using https://github.com/snok/install-poetry  action, like:
```yaml
jobs:
  my-job:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-python@v2
        with:
          python-version: 3.9
      - uses: snok/install-poetry@v1
        with:
          version: 1.2.0
      - name: Install dependencies
         run: poetry install
      - name: Run my script
        run: |
            poetry run python src/main.py
```

**Reply by junefish** [reply]
this is what I currently have:

```
jobs:
  build:
    strategy:
      fail-fast: false
      matrix:
        python-version: [3.10.6]
        poetry-version: [1.1.12]

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2

    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}

    - name: Install poetry ${{ matrix.poetry-version }}
      run: |
        python -m ensurepip
        python -m pip install --upgrade pip
        python -m pip install poetry==${{ matrix.poetry-version }}
    - name: View poetry --help
      run: poetry --help

    - name: Install dependencies
      shell: bash
      run: python -m poetry install --no-root
```

That all runs successfully, it only fails on the run (though of course that doesn't necessarily prove I'm installing it correctly).

**Reply by antonblr** [reply]
@junefish - I don't think you installing poetry correctly. Try workflow I posted above (adopted to your needs), or see official docs on install recommendations - https://python-poetry.org/docs/#ci-recommendations

**Reply by junefish** [reply]
I changed the poetry installation command:

```
-    - name: Install poetry ${{ matrix.poetry-version }}
-        run: |
-           python -m ensurepip
-           python -m pip install --upgrade pip
-           python -m pip install poetry==${{ matrix.poetry-version }}
+    - name: Install Poetry
+       uses: snok/install-poetry@v1
```

Now the dependency install is failing (previously successful):
```
Run python -m poetry install --no-root
/opt/hostedtoolcache/Python/3.10.6/x64/bin/python: No module named poetry
Error: Process completed with exit code 1.
```
I tried looking at https://python-poetry.org/docs/#ci-recommendations but I am new to Github Actions so I'm not sure how to adapt the instructions compared to a local install.

**Reply by antonblr** [reply]
@junefish - please follow my example -> use `run: poetry install --no-root` instead of `run: python -m poetry install --no-root` to install dependencies (`shell: bash` is redundant as well)

**Reply by junefish** [reply]
I updated as per your advice:
```
      - name: Check Poetry installation
        run: poetry --version

      - name: Install dependencies
        run: poetry install --no-root
```
Now the dependency install succeeds again, and it says that it can't find module config (which is another file in the same folder as main.py)
```

Run poetry run python -m src.main
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/todoist-habitica-sync/todoist-habitica-sync/src/main.py", line 8, in <module>
    from config import TODOIST_PRIORITY_TO_HABITICA_DIFFICULTY, get_settings
ModuleNotFoundError: No module named 'config'
Error: Process completed with exit code 1.
```
Running this exact code locally works, so I'm confused why Github Actions is throwing this error.

[Full YAML file](https://pastebin.com/JSuTDW5a)

**Reply by antonblr** [reply]
Hey @junefish - are you trying to run this https://github.com/junefish/todoist-habitica-sync/blob/main/src/main.py ?

You might want to set pythonpath, or/and add `__init__.py` files. Try this in your yaml first:

```yaml

- name: run sync
  run: |
    poetry run python src/main.py  # <---- slightly different command
  env:
    PYTHONPATH: "src"  # <---- added env var
    TODOIST_USER_ID: ${{ secrets.TODOIST_USER_ID }}
    TODOIST_API_KEY: ${{ secrets.TODOIST_API_KEY }}
    HABITICA_USER_ID: ${{ secrets.HABITICA_USER_ID }}
    HABITICA_API_KEY: ${{ secrets.HABITICA_API_KEY }}
```

**Reply by junefish** [reply]
That fixed it!!!!!

Well, I'm now getting a _different_ error, but it's a permissions error for Habitica, so I'll troubleshoot the API key that I set.  But the program is definitely running now, as it pulled the Todoist task that I created.  Thank you so much!

```
2022-10-18 00:26:13,384 (DelayTimer) [INFO]: Next check in 60 seconds.
2022-10-18 00:27:13,513 (TasksSync) [INFO]: New task github test, HABITICA_NEW
Error: -18 00:27:13,632 (TasksSync) [ERROR]: Unexpected network error: 401 Client Error: Unauthorized for url: https://habitica.com/api/v3/tasks/user
2022-10-18 00:27:13,632 (DelayTimer) [INFO]: Next check in 60 seconds.
```
